pkg/store/bucket.go

```diff
@@ -1352,150 +1405,141 @@ func (r *bucketIndexReader) ExpandedPostings(ms []*labels.Matcher) ([]uint64, er
 	return ps, nil
 }
 
-type postingGroup struct {
-	keys     labels.Labels
-	postings []index.Postings
+func getAllPostingsKeyLabel() labels.Label {
+	name, value := index.AllPostingsKey()
+	return labels.Label{Name: name, Value: value}
+}
 
-	aggregate func(postings []index.Postings) index.Postings
+// Logical result of each individual group is:
+// If addAll is set: ALL postings minus postings for removeKeys labels. (No need to merge postings for addKeys in this case)
+// If addAll is not set: Merge of postings for "addKeys" labels minus postings for removeKeys labels
+// This happens in ExpandedPostings.
+type postingGroup struct {
+	addAll     bool
+	addKeys    []labels.Label
+	removeKeys []labels.Label
 }
 
-func newPostingGroup(keys labels.Labels, aggr func(postings []index.Postings) index.Postings) *postingGroup {
+func newPostingGroup(addAll bool, addKeys, removeKeys []labels.Label) *postingGroup {
```

Comment:
Also we can do the same for nothing matches:

`var NoMatchingGroup = &postingGroup{}`

This work as the ptr addresses will be different and ptr is just int so it's comparable.

Reply:
I'm fine with using `noMatchingGroup` as sentinel value (checked for in Expanded_Postings), but `allPostingsGroup` (why Exported?) should be `var allPostingsGroup  = newPostingGroup(true, nil, nil)` to avoid extra checks for it.

Reply:
👍 Added `allPostingsGroup` and `emptyPostingsGroup` vars.

Reply:
Thanks - yea no need to export them.